### PR TITLE
do not enable placement rules automatically

### LINKED ISSUE #2219: PD placement rules should be enabled if using TiFlash
## Bug Report

**What version of TiDB Operator are you using?**

latest release

**What did you do?**

Start TiDBCluster with TiFlash

**What did you expect to see?**

TiFlash and TiDBcluster running normally

**What did you see instead?**

TiFlash failed to start because the PD placement-rules is not enabled, this is a prerequisite for TiFlash. Error log of TiFlash

```
klo yeya24-tiflash-0 -c proxylog
[2020/04/16 19:27:56.857 +00:00] [ERROR] [util.rs:327] ["request failed"] [err="Grpc(RpcFailure(RpcStatus { status: FailedPrecondition, details: Some(\"placement rules is disabled\") }))"]
[2020/04/16 19:27:56.857 +00:00] [ERROR] [util.rs:327] ["request failed"] [err="Grpc(RpcFailure(RpcStatus { status: FailedPrecondition, details: Some(\"placement rules is disabled\") }))"]
[2020/04/16 19:27:56.857 +00:00] [FATAL] [server.rs:287] ["failed to start node: Other(\"[src/pd/util.rs:335]: fail to request\")"]
```

PD config status:

```
pdctl config show | grep enable-placement
    "enable-placement-rules": "false",
```

If I manually enabled replacement rules in PD, then TiFlash starts perfectly

### PR DESCRIPTION
**Pull request by DanielZhangQD**
### What problem does this PR solve? 
Fix #2219 

### What is changed and how does it work?
If users enable TiFlash for the existing TiDB Cluster, enable placement rules automatically with pdapi.
### Check List 

Tests 

 - Manual test (add detailed scripts or steps below)
   - Enable TiFlash for the existing TiDB Cluster, enable-placement-rules is set automatically and PD cluster will not be rolling updated.
   - Create cluster with enable-placement-rules set, tiflash will start successfully and the operator will not update the enable-placement-rules config.

Code changes

 - Has Go code change

Related changes

 - Need to cherry-pick to the release branch

### Does this PR introduce a user-facing change?:

```release-note
Set `enable-placement-rules` to `true` for PD if TiFlash is enabled in the cluster
```

### CODE REVIEW

**Comment by yeya24** [comment]
So we still have to enable this on our own? Instead of managing by operator?

**Comment by DanielZhangQD** [comment]
> So we still have to enable this on our own? Instead of be managed by the operator?

For an existing cluster, users have to enable it manually via pd-ctl, update the configuration file does not take effect.

File: pkg/manager/member/pd_member_manager.go [hunk 1]
```diff
@@ -741,17 +741,7 @@ func getPDConfigMap(tc *v1alpha1.TidbCluster) (*corev1.ConfigMap, error) {
 		config.Dashboard.TiDBCertPath = path.Join(tidbClientCertPath, corev1.TLSCertKey)
 		config.Dashboard.TiDBKeyPath = path.Join(tidbClientCertPath, corev1.TLSPrivateKeyKey)
 	}
-	// TiFlash requires PD to enable the `replication.enable-placement-rules`
-	// Check detail in https://pingcap.com/docs/stable/reference/tiflash/deploy/
-	if tc.Spec.TiFlash != nil {
```
Comment (weekface) [inline pkg/manager/member/pd_member_manager.go:746]: Enable placement rules if we create the cluster?

Comment (DanielZhangQD) [reply pkg/manager/member/pd_member_manager.go:746]: When creating the configmap, it does not know whether the cluster is new or not, and it should not care about it.
I think we can make it clear in the doc, after all, users also have to add the TiFlash configurations, the TiFlash docs also request users to enable it manually.

Comment (Yisaer) [reply pkg/manager/member/pd_member_manager.go:746]: Currently there are 2 steps for user to enable Tiflash. First, enable rule-placement by pd-ctl. Second configure Tiflash Table Replicas by sql. I think only SQL configuration should be exposed to the users excluding pdapi. We have some seniors that users couldn't get in touch with pdapi. 

As TiFlash is [supprted to be enabled in an existed cluster](https://pingcap.com/docs-cn/stable/reference/tiflash/deploy/#%E5%9C%A8%E5%8E%9F%E6%9C%89-tidb-%E9%9B%86%E7%BE%A4%E4%B8%8A%E6%96%B0%E5%A2%9E-tiflash-%E7%BB%84%E4%BB%B6), I think we could enabled it `placement rule` by pdapi before syncing tiflash.

Comment (DanielZhangQD) [reply pkg/manager/member/pd_member_manager.go:746]: updated

**Comment by DanielZhangQD** [comment]
@Yisaer @weekface comments addressed, PTAL.

File: pkg/pdapi/pdapi.go [hunk 2]
```diff
@@ -512,6 +515,24 @@ func (pc *pdClient) SetStoreLabels(storeID uint64, labels map[string]string) (bo
...
+	}
+	res, err := pc.httpClient.Post(apiURL, "application/json", bytes.NewBuffer(data))
+	if err != nil {
+		return err
+	}
+	defer httputil.DeferClose(res.Body)
+	if res.StatusCode == http.StatusOK {
+		return nil
+	}
+	err2 := httputil.ReadErrorBody(res.Body)
```
Comment (Yisaer) [inline pkg/pdapi/pdapi.go:532]: why not directly use err here

File: pkg/pdapi/pdapi.go [hunk 3]
```diff
@@ -697,6 +718,7 @@ const (
 	DeleteMemberByIDActionType         ActionType = "DeleteMemberByID"
 	DeleteMemberActionType             ActionType = "DeleteMember "
 	SetStoreLabelsActionType           ActionType = "SetStoreLabels"
+	UpdatePlacementRulesActionType     ActionType = "UpdateEnablePlacementRules"
```
Comment (Yisaer) [inline pkg/pdapi/pdapi.go:721]: I think PlacementRules is part of the `ReplicationConfig`, We should name it as `SetReplicationConfig`

Comment (DanielZhangQD) [reply pkg/pdapi/pdapi.go:721]: There are other fields in this struct, here I only update this one field, if need to update other fields, we should define another function, they may be used in different scenarios.

Comment (DanielZhangQD) [reply pkg/pdapi/pdapi.go]: updated.

File: pkg/pdapi/pdapi.go (hunk 2 above)
Comment (DanielZhangQD) [reply pkg/pdapi/pdapi.go]: updated

**Pull request merged**